fs/dynchunksize/dynchunksize.go

```diff
@@ -0,0 +1,37 @@
+// Dynamically calculate suitable chunk size for large uploads
+package dynchunksize
+
+import (
+	"github.com/rclone/rclone/fs"
+)
+
+const oneMiB = int64(1024 * 1024)
+
+/*
+ Compute the minimum chunk size needed to fit within the given maximum number of parts, rounded up to the nearest MiB
+
+ For most backends, (chunk_size) * (concurrent_upload_routines) memory will be required so we want to use the smallest
+ possible chunk size that's going to allow the upload to proceed. Rounding up to the nearest MiB on the assumption
+ that some backends may only allow integer type parameters when specifying the chunk size.
+
+ Returns the default chunk size if it is sufficiently large enough to support the given file size otherwise returns the
+ smallest chunk size necessary to allow the upload to proceed.
+*/
+func ComputeChunkSize(fileSize fs.SizeSuffix, maxParts int, defaultChunkSize fs.SizeSuffix) fs.SizeSuffix {
+	requiredChunks := fileSize / defaultChunkSize
+	if requiredChunks < fs.SizeSuffix(maxParts) || (requiredChunks == fs.SizeSuffix(maxParts) && fileSize%defaultChunkSize == 0) {
+		fs.Debugf(nil, "size: %v, parts: %v, default: %v; returned default value", fileSize, maxParts, defaultChunkSize)
+		return defaultChunkSize
+	}
+
+	minChunk := fileSize / fs.SizeSuffix(maxParts)
+	remainder := minChunk % fs.SizeSuffix(oneMiB)
+	if remainder != 0 {
+		minChunk += fs.SizeSuffix(oneMiB) - remainder
+	} else if int64(fileSize/minChunk) == int64(maxParts) && int64(fileSize)%int64(maxParts) != 0 { // when right on the boundary, we need to add a MiB
+		minChunk += fs.SizeSuffix(oneMiB)
+	}
+
+	fs.Debugf(nil, "size: %v, parts: %v, default: %v, new: %v; default chunk size insufficient, returned new chunk size", fileSize, maxParts, defaultChunkSize, minChunk)
```

Comment:
Not having the object to log here will make these logs not very useful. So either pass the object to log in, or comment out the Debug.